### ISSUE TITLE
VersionSpec wildcard problem

### ISSUE THREAD

**Issue by mcg1969** [conda/conda#2485]
I don't think `numpy 1.1*` should match `numpy 1.10`, but that's exactly what happens:
```
from conda.version import VersionSpec
vspec = VersionSpec('1.1*')
vspec.match('1.10')
```
This returns `True` but it should be `False`, in my view. 

I think it has been this way as long as the `VersionSpec` object has been around (I bisected to 3.14).

The reason for this is that `VersionSpec` converts `1.1*` to the regex `(1\.1.*)$`. (As far as I can tell the grouping is superfluous.) I think it should do something like this instead: `1\.1(?:\..*)?$`

Thoughts? @ukoethe

**Comment by mcg1969** [comment]
I think we can add something like this to the regex branch of `VersionSpec.__new__`:
```
            if rx.endswith('.*') and not rx.endswith('..*'):
                rx = rx[:-2] + r'(?:\..*)?'
```
The extra test makes sure we don't add an extra period if the spec already had one.

**Comment by mcg1969** [comment]
Maybe something like this is better:
```
            # endwild_re = re.compile(r'.*[a-zA-Z0-9]\.\*$')
            if endwild_re.match(rx):
                rx = rx[:-2] + r'(?:\..*)?'
```

**Comment by mcg1969** [comment]
Hmm, this doesn't work with `1.1post1`...

**Comment by mcg1969** [comment]
I think that the proper resolution here for the _trailing_ wildcard is to _eliminate the use of regexes_ and rely on `VersionOrder` instead. So, for instance, `2.7*` matches any normalized version with 2 and 7 in the first two "slots". In this case, it's equivalent to `>=2.7,<2.8`, but we can't always translate to numerical inequalities, because we might also want to handle cases like this: `2.7post*`. So it really is a special kind of match.

I'm unaware of any actual use of the wildcard in a _non-trailing_ position. In my view that should be deprecated altogether.

**Comment by mcg1969** [comment]
@kalefranz want to make sure you've seen this. @ukoethe is responsible for `VersionOrder`, which is of course an awesome piece of work. The problem here involves the one case where _we don't use it_.

**Comment by mcg1969** [comment]
We may not be able to eliminate the regex match altogether---specifically those cases where the version number can't be parsed by `VersionOrder`. Thankfully those situations are rare and can be discouraged.

**Comment by mcg1969** [comment]
I've created a `startswith` method for VersionOrder, and changed `VersionSpec` to use it only in the very simple case when only a trailing wildcard is found. In all other wildcard cases, the regex is still used.

Interestingly, [PEP440](https://www.python.org/dev/peps/pep-0440/) has the notion of a "prefix match", and this implementation is reasonably close to it. And the `VersionOrder` code already accepts asterisks in its version numbering.

**Comment by kalefranz** [comment]
Will be included in `4.1.0` release.
